### DISCUSSION TITLE
Key Insight about num_workers and why your code may be running slow

### DISCUSSION THREAD

**Discussion by Finix-07** [mrdbourke/pytorch-deep-learning#1135] [Ideas]
When I first started working on data loaders, I noticed my training and testing times were much slower than what was demonstrated in the tutorial videos. Initially, I assumed this was due to a bottleneck in my system, but the same issue persisted even on faster machines I tested.
![image](https://github.com/user-attachments/assets/d1ef4922-665d-4917-a923-abac4f58f504)

Today, while running my code, I noticed my CPU usage spiking. My initial thought was that the model was somehow training on the CPU, but Task Manager confirmed the GPU was being used, which left me puzzled. I had previously believed that increasing the `num_workers` parameter would speed up the data loading process, thinking more workers would result in faster batch transfers to the GPU. However, through experimentation, I discovered that **_higher `num_workers` actually put significant strain on the CPU_**.
![image](https://github.com/user-attachments/assets/cef5e063-766a-413a-a740-4fe20040fc58)

The breakthrough came when I compared the results on my local system (with **_12 CPU cores_**) to Google Colab (with only **_2 cores_**). By adjusting `num_workers` to **_2_** on my local system, I saw a **_drastic reduction_** in both training time and CPU usage. This experiment revealed a key insight: when working with smaller datasets, a high `num_workers` value can become a bottleneck. **_Setting `num_workers` to `0` resulted in significantly faster training times_**.

NUM_WORKERS = 2
![image](https://github.com/user-attachments/assets/28a7ad21-6d98-402e-960c-68407582f0e6)
NUM_WORKERS = 0
![image](https://github.com/user-attachments/assets/f5bd7ddf-8529-4554-8c55-fe95d490977c)
CPU UTIL DURING THESE 2 
![image](https://github.com/user-attachments/assets/6d93d6a2-10c2-4d4b-9ffe-129abe723bd0)

### **_Conclusion_**:
For small datasets, keeping **_`num_workers` low_** or even at `0` can drastically improve performance by minimizing CPU load and optimizing data transfer to the GPU.

Edit: added pics

**Comment by LuluW8071** [comment]
That also might be the case. And, things to note is that, data loading and processing also depends on ur disk speed (for instance, ssd is faster than hdd to load the data) and gpu specs

Num workers doesn't necessarily load the data faster. It justs prepare the batches of data in parallel to make it readily available to transfer into gpu, so it doesn't have to wait for additional time. As, colab drive is slower comparatively to other platforms. This also might be the case where setting `num_workers` to 0 or 2 doesn't show much improvement.

Furthermore, in colab if u didn't restart the gpu and trained again while not freeing the previous model instance running on gpu. It may have caused slow training process.

>[!TIP]
>To speed of ur training by >1.5x process try `torch.compile`.
> However, if your model is reliant on RNNs type complie approach slows down training process.

```py
model = your_model_class( )
compiled_model = torch.compile(model)

# Train using compiled_model
```

**Reply by Finix-07** [reply]
`torch.compile` is indeed really great..... amazing tip !!
